src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/DeviceTest.cs

```diff
@@ -62,47 +62,46 @@ protected static bool MonitorAdbLogcat (Func<string, bool> action, int timeout =
 				CreateNoWindow = true,
 				WindowStyle = ProcessWindowStyle.Hidden,
 			};
+
+			bool didActionSucceed = false;
 			using (var proc = Process.Start (info)) {
-				var sw = new Stopwatch ();
-				try {
-					TimeSpan time = TimeSpan.FromSeconds (timeout);
-					while (time.TotalMilliseconds > 0) {
-						sw.Start ();
-						if (action (proc.StandardOutput.ReadLine ()))
-							return true;
-						time = time.Subtract (TimeSpan.FromMilliseconds (sw.ElapsedMilliseconds));
-						sw.Reset ();
+				var sb = new StringBuilder ();
+
+				proc.OutputDataReceived += (sender, e) => {
+					if (e.Data != null) {
+						sb.AppendLine (e.Data);
+						if (action (e.Data)) {
+							didActionSucceed = true;
+							proc.Kill ();
+						}
 					}
-				} finally {
-					sw.Stop ();
+				};
+
+				proc.BeginOutputReadLine ();
+				if (!proc.WaitForExit (timeout * 1000)) {
 					proc.Kill ();
-					proc.WaitForExit ();
 				}
-				return false;
+
+				File.WriteAllText (logcatFilePath, sb.ToString ());
```

Comment:
So this changes the behavior to timeout at 15 seconds overall, instead of 10 seconds of no `adb logcat` output. I think that is OK if the tests are still green.

Could we instead use `File.CreateText` to get a `StreamWriter`, and call `StreamWriter.WriteLine()` for each line? Instead of holding the entire log in memory in a `StringBuilder`. This might prevent some OOM issues down the road.